demos/classification_benchmark_demo/cpp_gapi/main.cpp

```diff
@@ -0,0 +1,549 @@
+    if (name_endline_pos == std::string::npos) {
+        throw std::runtime_error("Cannot parse CfgParam from: " + line +
+                                 "\nExpected separator \":\"");
+    }
+
+    std::string name = line.substr(0, name_endline_pos);
+    std::string value = line.substr(name_endline_pos + 1);
+
+    return cv::gapi::wip::onevpl::CfgParam::create(name, value,
+                                                   /* vpp params strongly optional */
+                                                   name.find("vpp.") == std::string::npos);
+}
+
+static std::vector<cv::gapi::wip::onevpl::CfgParam> parseVPLParams(const std::string& cfg_params) {
+    std::vector<cv::gapi::wip::onevpl::CfgParam> source_cfgs;
+    std::stringstream params_list(cfg_params);
+    std::string line;
+    while (std::getline(params_list, line, ',')) {
+        source_cfgs.push_back(createFromString(line));
+    }
+    return source_cfgs;
+}
+
+static cv::gapi::GKernelPackage getKernelPackage(const std::string& type) {
+    if (type == "opencv") {
+        return cv::gapi::combine(cv::gapi::core::cpu::kernels(), cv::gapi::imgproc::cpu::kernels());
+    } else if (type == "fluid") {
+        return cv::gapi::combine(cv::gapi::core::fluid::kernels(), cv::gapi::imgproc::fluid::kernels());
+    } else {
+        throw std::logic_error("Unsupported kernel package type: " + type);
+    }
+    GAPI_Assert(false && "Unreachable code!");
+}
+}  // namespace util
+
+namespace nets {
+G_API_NET(Classification, <cv::GMat(cv::GMat)>, "classification");
+}
+
+struct IndexScore {
+    using IndicesStorage = std::list<size_t>;
+    using ConfidenceMap = std::multimap<float, typename IndicesStorage::iterator>;
+
+    using ClassDescription = std::tuple<float, size_t, std::string>;
+    using LabelsStorage = std::vector<ClassDescription>;
+
+    IndexScore() = default;
+
+    void getScoredLabels(const std::vector<std::string> &in_labes,
+                         LabelsStorage &out_scored_labels_to_append) const {
+        try {
+            // fill starting from max confidence
+            for (auto conf_index_it = max_confidence_with_indices.rbegin();
+                 conf_index_it != max_confidence_with_indices.rend();
+                 ++conf_index_it) {
+                std::string str_label = in_labes.at(*conf_index_it->second);
+                out_scored_labels_to_append.emplace_back(conf_index_it->first, *conf_index_it->second,  std::move(str_label));
+            }
+        } catch (const std::out_of_range& ex) {
+            throw std::out_of_range(std::string("Provided labels file doesn't contain detected class index\nException: ") + ex.what());
+        }
+    }
+
+    static IndexScore create_from_array(const float *out_blob_data_ptr, size_t out_blob_element_count,
+                                        size_t top_k_amount) {
+        IndexScore ret;
+        if (!out_blob_data_ptr) {
+            return IndexScore();
```

Comment:
The question is not about `IndexScore`, but about `out_blob_data_ptr` being null. Why do you care? If G-API provides nullptr as `out_blob.ptr<float>()`, there's probably a bug in it, so you shouldn't pretend that it's fine

Reply:
because `IndexScore` here doesn't know anything about G-API and high-level rules and might be used separately (after refactoring for example)  - its responsibility is build data structure from input.

So this check on empty array must be done in G-API kernel - I think I had missed that

Reply:
I'm ok with the current state. Now I'm just trying to understand:
1. How do you define high-level rules.
2. How do you define class having knowledge about something. And about G-API and high-level rules in particular. I would guess mentioning any G-API specific entity in scope of `IndexScore` makes it to know about G-API. But I can't apply the same idea to "rules".
3. > after refactoring for example

    Can you name an event that would trigger such refactoring?